numpy/core/src/multiarray/datetime.c

```diff
@@ -2852,85 +2852,223 @@ convert_datetime_to_pyobject(npy_datetime dt, PyArray_DatetimeMetaData *meta)
+            out->ps  = (npy_int32)extract_unit_64(&td, 1000LL*1000);
+            out->as  = (npy_int32)td;
+            break;
+        default:
+            PyErr_SetString(PyExc_RuntimeError,
+                            "NumPy timedelta metadata is corrupted with invalid "
+                            "base unit");
+            return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * Converts a timedelta into a PyObject *.
+ *
+ * Not-a-time is returned as the string "NaT".
+ * For microseconds or coarser, returns a datetime.timedelta.
+ * For units finer than microseconds, returns an integer.
+ */
+NPY_NO_EXPORT PyObject *
+convert_timedelta_to_pyobject(npy_timedelta td, PyArray_DatetimeMetaData *meta)
+{
+    npy_timedeltastruct tds;
+
+    /*
+     * Convert NaT (not-a-time) into None.
+     */
+    if (td == NPY_DATETIME_NAT) {
+        Py_RETURN_NONE;
+    }
+
+    /*
+     * If the type's precision is greater than microseconds, is
+     * Y/M/B (nonlinear units), or is generic units, return an int
+     */
+    if (meta->base > NPY_FR_us ||
+                    meta->base == NPY_FR_Y ||
+                    meta->base == NPY_FR_M ||
+                    meta->base == NPY_FR_GENERIC) {
+        return PyLong_FromLongLong(td);
     }
+
+    if (convert_timedelta_to_timedeltastruct(meta, td, &tds) < 0) {
+        return NULL;
+    }
+
     /*
      * If it would overflow the datetime.timedelta days, return a raw int
      */
-    if (days < -999999999 || days > 999999999) {
+    if (tds.day < -999999999 || tds.day > 999999999) {
         return PyLong_FromLongLong(td);
     }
     else {
-        return PyDelta_FromDSU(days, seconds, useconds);
+        return PyDelta_FromDSU(tds.day, tds.sec, tds.us);
+    }
+}
+
+/*
+ * We require that if d is a PyDelta, then
+ * hash(numpy.timedelta64(d)) == hash(d).
+ * Where possible, convert dt to a PyDelta and hash it.
+ */
+NPY_NO_EXPORT npy_hash_t
+timedelta_hash(PyArray_DatetimeMetaData *meta, npy_timedelta td)
+{
+    PyObject *obj;
+    npy_hash_t res = 0;
+    npy_timedeltastruct tds;
+
+    if (td == NPY_DATETIME_NAT || meta->base == NPY_FR_GENERIC) {
+        return -2;
+    }
+
+    /* Y and M can be converted to each other but not to other units */
+
+    if (meta->base == NPY_FR_Y) {
+        res = (npy_hash_t)(td * 12);
+        return (res == -1) ? -2 : res;  /* avoid -1 (error) */
     }
+    if (meta->base == NPY_FR_M) {
+        res = (npy_hash_t)td;
+        return (res == -1) ? -2 : res;  /* avoid -1 (error) */
+    }
+
+    if (convert_timedelta_to_timedeltastruct(meta, td, &tds) < 0) {
+        return -1;
+    }
+
+    if (tds.day < -999999999 || tds.day > 999999999) {
+        obj = PyBytes_FromStringAndSize((const char *)&tds, sizeof(tds));
+    } else {
+        res = tds.ps * 3 + tds.as;
+
+        obj = PyDelta_FromDSU(tds.day, tds.sec, tds.us);
+    }
+
+    if (obj == NULL) {
+        return -1;
+    }
+
+    res += PyObject_Hash(obj);
```

Comment:
Missing check for -1 here to

Reply:
Eh? it's on line 3071.

Reply:
I assume from the deleted comment that you worked out what I meant:)

Reply:
Yes. Sorry :-)